storage/repair.go

```diff
@@ -0,0 +1,278 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package storage
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3db/client"
+	"github.com/m3db/m3db/clock"
+	"github.com/m3db/m3db/retention"
+	"github.com/m3db/m3db/storage/block"
+	"github.com/m3db/m3db/storage/repair"
+	"github.com/m3db/m3db/topology"
+	"github.com/m3db/m3x/errors"
+	"github.com/m3db/m3x/log"
+
+	"github.com/uber-go/tally"
+)
+
+var (
+	errNoRepairOptions = errors.New("no repair options")
+)
+
+type recordFn func(namespace string, shard databaseShard, diffRes repair.MetadataComparisonResult)
+
+type shardRepairer struct {
+	opts    Options
+	rtOpts  retention.Options
+	session client.AdminSession
+
+	recordFn  recordFn
+	logger    xlog.Logger
+	scope     tally.Scope
+	nowFn     clock.NowFn
+	blockSize time.Duration
+	origin    topology.Host
+	replicas  int
+}
+
+func newShardRepairer(opts Options, rpOpts repair.Options) (databaseShardRepairer, error) {
+	newSessionFn := rpOpts.NewAdminSessionFn()
+	session, err := newSessionFn()
+	if err != nil {
+		return nil, err
+	}
+
+	iOpts := opts.InstrumentOptions()
+	rtOpts := opts.RetentionOptions()
```

Comment:
For these short names, I've actually not been camel casing them. i.e. `iopts := opts.InstrumentOptions()` and `ropts := opts.RetentionOptions()`